src/Integration/Binding/JsTsProjectTypeIndicator.cs

```diff
@@ -0,0 +1,115 @@
+
+namespace SonarLint.VisualStudio.Integration.Binding
+{
+    [Export(typeof(IJsTsProjectTypeIndicator))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class JsTsProjectTypeIndicator : IJsTsProjectTypeIndicator
+    {
+        readonly ISonarLanguageRecognizer sonarLanguageRecognizer;
+        readonly IFolderWorkspaceService folderWorkspaceService;
+        readonly IFileSystem fileSystem;
+
+        [ImportingConstructor]
+        public JsTsProjectTypeIndicator(ISonarLanguageRecognizer sonarLanguageRecognizer, IFolderWorkspaceService folderWorkspaceService)
+            : this(sonarLanguageRecognizer, folderWorkspaceService, new FileSystem())
+        {
+
+        }
+
+        internal JsTsProjectTypeIndicator(ISonarLanguageRecognizer sonarLanguageRecognizer, IFolderWorkspaceService folderWorkspaceService, IFileSystem fileSystem)
+        {
+            this.sonarLanguageRecognizer = sonarLanguageRecognizer;
+            this.folderWorkspaceService = folderWorkspaceService;
+            this.fileSystem = fileSystem;
+        }
+        
+        public bool IsJsTs(Project dteProject)
+        {
+            //When opened as folder there can be a dteProject if a file is open
+            //If there is a dteProject and it's opened as a folder
+            //Folder search takes precedense for consistency
+            if (folderWorkspaceService.IsFolderWorkspace())
+            {
+                return HasJsTsFileOnDisk();
+            }
+            else
+            {
+                Debug.Assert(dteProject != null, "When it's not folder workspace we expect dteProject not to be null");
+
+                return HasJsTsFileInProject(dteProject.ProjectItems);
+            }            
+        }
+
+        private bool HasJsTsFileOnDisk()
+        {
+            string root = folderWorkspaceService.FindRootDirectory();
+
+            var fileList = fileSystem.Directory.EnumerateFiles(root, "*", SearchOption.AllDirectories).Where(x => !x.Contains("\\node_modules\\"));
+
+            foreach (var file in fileList)
+            {
+                if(IsFileJsTs(file))
+                {
+                    return true;
+                }    
+            }
+            return false;
+        }
+
+        private bool IsFileJsTs(string fileName)
+        {
+            var analysisLanguage = sonarLanguageRecognizer.GetAnalysisLanguageFromExtension(fileName);
+            return analysisLanguage == Core.Analysis.AnalysisLanguage.Javascript || analysisLanguage == Core.Analysis.AnalysisLanguage.TypeScript;
+        }
+
+        private bool HasJsTsFileInProject(ProjectItems projectItems)
+        {
+
+            foreach (ProjectItem item in projectItems)
+            {
+                if (IsFileJsTs(item.Name))
+                {
+                    return true;
+                }
+                if (item.ProjectItems?.Count > 0)
+                {
+                    if (HasJsTsFileInProject(item.ProjectItems))
```

Comment:
SQ raised a code smell for this nested `if` - could be a single `if`.

Reply:
I have seen it but I think it would reduce the readability. We go to the inner loop only if the item has child items and if we merge it, this won't be obvious.